bundler/lib/bundler/rubygems_ext.rb

```diff
@@ -61,16 +61,97 @@ def open_file_with_lock(path, &block)
   require "rubygems/platform"
 
   class Platform
-    JAVA  = Gem::Platform.new("java")
-    MSWIN = Gem::Platform.new("mswin32")
-    MSWIN64 = Gem::Platform.new("mswin64")
-    MINGW = Gem::Platform.new("x86-mingw32")
-    X64_MINGW = [Gem::Platform.new("x64-mingw32"),
-                 Gem::Platform.new("x64-mingw-ucrt")].freeze
-    UNIVERSAL_MINGW = Gem::Platform.new("universal-mingw")
-    WINDOWS = [MSWIN, MSWIN64, UNIVERSAL_MINGW].flatten.freeze
-    X64_LINUX = Gem::Platform.new("x86_64-linux")
-    X64_LINUX_MUSL = Gem::Platform.new("x86_64-linux-musl")
+    unless respond_to?(:generic)
+      JAVA  = Gem::Platform.new("java")
+      MSWIN = Gem::Platform.new("mswin32")
+      MSWIN64 = Gem::Platform.new("mswin64")
+      MINGW = Gem::Platform.new("x86-mingw32")
+      X64_MINGW = [Gem::Platform.new("x64-mingw32"),
+                   Gem::Platform.new("x64-mingw-ucrt")].freeze
+      UNIVERSAL_MINGW = Gem::Platform.new("universal-mingw")
+      WINDOWS = [MSWIN, MSWIN64, UNIVERSAL_MINGW].flatten(1).freeze
+      X64_LINUX = Gem::Platform.new("x86_64-linux")
+      X64_LINUX_MUSL = Gem::Platform.new("x86_64-linux-musl")
+
+      GENERICS = [RUBY, JAVA, *WINDOWS].freeze
+      private_constant :GENERICS
+
+      GENERIC_CACHE = GENERICS.each_with_object({}) {|g, h| h[g] = g }
+      private_constant :GENERIC_CACHE
+
+      def generic
+        GENERIC_CACHE[self] ||= begin
+          found = GENERICS.find do |match|
+            self === match
+          end
+          found || Gem::Platform::RUBY
+        end
+      end
+
+      class << self
+        def generic(platform)
+          return Gem::Platform::RUBY if platform.nil? || platform == Gem::Platform::RUBY
+          platform.generic
+        end
+
+        def platform_specificity_match(spec_platform, user_platform)
+          return -1 if spec_platform == user_platform
+          return 1_000_000 if spec_platform.nil? || spec_platform == Gem::Platform::RUBY || user_platform == Gem::Platform::RUBY
+
+          os_match = if spec_platform.os == user_platform.os
+            0
+          else
+            1
+          end
+          cpu_match = if spec_platform.cpu == user_platform.cpu
+            0
+          elsif spec_platform.cpu == "arm" && user_platform.cpu.to_s.start_with?("arm")
+            0
+          elsif spec_platform.cpu.nil? || spec_platform.cpu == "universal"
+            1
+          else
+            2
+          end
+          version_match = if spec_platform.version == user_platform.version
+            0
+          elsif spec_platform.version.nil?
+            1
+          else
+            2
+          end
```

Comment:
Just reviewing the extra changes in the implementation on top of moving the methods around, what's the reason for inlining these methods into local variables?

Reply:
original goal was to avoid introducing new methods into `Gem::Platform` that were not necessary

Reply:
I see. I think as long as they are private it'd be ok, but it's fine if you prefer it this way.